src/MemStore.cc

```diff
@@ -526,6 +544,9 @@ MemStore::copyFromShm(StoreEntry &e, const sfileno index, const Ipc::StoreMapAnc
     debugs(20, 5, "mem-loaded all " << e.mem_obj->endOffset() << '/' <<
            anchor.basics.swap_file_sz << " bytes of " << e);
 
+    if (e.mem().adjustableBaseReply().pstate != Http::Message::psParsed)
```

Comment:
Should not we also check that it is Http::Message::psParsed before returning true in the !wasEof block above?

Reply:
I do not think so: MemStore::get() -- our caller -- does not promise a StoreEntry object with parsed headers. A true MemStore::copyFromShm() results simply means "no errors". It does not imply "we've got headers". In a hypothetical case, the shared memory may have no HTTP bytes to copy/parse yet (but the entry itself may already exist).

This usage/semantics seems similar to a Disks::get() method that returns an entry before loading any HTTP response headers. See Store::Controller::peek() for an example of how these get() methods are used in the same context.

We _could_ require shared memory to provide headers on get(), but I am not sure adding this difference with Disks::get() is a good idea, and the change itself feels out of this PR scope.

Reply:
> our caller -- does not promise a StoreEntry object with parsed headers

However we start writing to MemStore when there is an entry with (parsed) headers,via Client::setFinalReply()-> StoreEntry::startWriting() path. Probably there is another way how some initial (unparsed) bytes of a response could be written to MemStore (and subsequently returned via MemStore::get()).

Reply:
> we start writing to MemStore when there is an entry with (parsed) headers,via Client::setFinalReply()-> StoreEntry::startWriting() path

The code you are thinking about may be running in another worker process, so we would have to go down to low-level locking/appending events to be sure that the headers are always there when MemStore::get() might find a _readable_ entry being written by _another_ kid.

* I am not certain, but it looks like that MemStore::startCaching() does create readable shared memory entries that have no HTTP bytes (yet). That moment is enough to violate the "a readable shared memory entry implies presence of full HTTP headers" invariant.
* After that moment, one kid could be appending a few HTTP response header bytes to shared memory while the other kid is reading them... Plenty of opportunities for that invariant violations during this process!

> Probably there is another way how some initial (unparsed) bytes of a response could be written to MemStore (and subsequently returned via MemStore::get()).

And for the proposed "does not promise" defense to work, the number of those initial bytes can be zero! It is not the number of bytes that matters but the availability of a readable shared memory entry. That entry may be empty!

One might hope that startCaching() itself is not called until all response headers are written to local memory (from where MemStore would copy them to shared memory), but at least one StoreEntry::swapOut() caller does not seem to care about our `ENTRY_FWD_HDR_WAIT` hack, and MemStore itself does not care about that flag at all. AFAICT, StoreEntry::swapOut() can be called for an ENTRY_FWD_HDR_WAIT entry without any headers in local memory.

More importantly, there is no enforced invariant (AFAICT) that would preclude some (current or future) code from making a memory cache entry available for reading before it has full HTTP response headers. If there is such an invariant, then we should probably fix MemStore code, but I would argue that such a fix is best developed _after_ this PR because the problem (if any) is far from trivial and is mostly unrelated to this PR focus/changes.

FWIW, I see no such invariant in MemStore::startCaching() and copyToShm() code. As mentioned earlier, current MemStore writing code appears to not care about HTTP header boundary (in a sense related to this discussion) and higher-level swapOut() code might run quite early, before the headers are written to local memory.

Reply:
> MemStore::startCaching() does create readable shared memory entries that have no HTTP bytes (yet).

I checked the code and think that this is correct: though copyToShm() call follows startCaching(), these both operations are separate (not atomic) - another process may start reading between these two calls.